README.md
# connect-combo

Connect middleware for assets combo, resolve url like `http://127.0.0.1??a.js,b.js`.

[![Build Status](https://travis-ci.org/popomore/connect-combo.png?branch=master)](https://travis-ci.org/popomore/connect-combo)
[![Coverage Status](https://coveralls.io/repos/popomore/connect-combo/badge.png?branch=master)](https://coveralls.io/r/popomore/connect-combo?branch=master)

---

## Install

```
$ npm install connect-combo -g
```

## Usage

```
var app = connect()
  .use(combo())
  .listen(3000);
```

See [example](https://github.com/popomore/connect-combo/blob/master/examples/) 

## Option

- `directory` specify local directory base, default `process.cwd()`.

- `proxy` specify remote server, default `false`.

- `cache` cache file from remote server, default `false`.

- `log` show log on terminal, default `false`.

- `beforeProxy` handle request before send to proxy, default `null`.

## License

MIT


HISTORY.md

0.5.0 / 2016-06-29
==================

  * feat: carry http header from origin url if use combo as proxy only (#12)

0.4.0 / 2016-06-20
==================

  * feat: proxy support function (#11)

0.3.1 / 2015-10-20
==================

  * fix: proxy can be a url path

0.3.0 / 2015-09-25
==================

  * refactor: use mkdirp module
  * feat: options.proxy support array and can request multi url
  * fix: return buffer rather than string

## 0.2.7

rejectUnauthorized when request https

## 0.2.6

add beforeProxy option

## 0.2.5

fix combo single file

## 0.2.4

use urllib

## 0.2.3

refactor for url parse

## 0.2.2

fix content type in static options

## 0.2.1

directory support function

## 0.2.0

First commit


test/file.js
'use strict';

require('should');
var fs = require('fs');
var join = require('path').join;
var url = require('url');
var sinon = require('sinon');
var File = require('..').File;
var fixture = join(__dirname, 'fixture');

describe('File', function() {

  var options;

  beforeEach(function() {
    options = {
      directory: fixture,
      proxy: 'http://static.alipayobjects.com',
      cache: false,
    };
  });

  it('before proxy', function(done) {
    options.beforeProxy = function(urlPath, cb, next) {
      if (urlPath === 'beforeproxy.js') {
        this.emit('found');
        cb(null, 'beforeproxy');
      } else {
        next();
      }
    };

    var spy = sinon.spy();
    new File('beforeproxy.js', options)
      .on('found', spy)
      .end(function(err, data) {
        if (err) return done(err);
        data.should.be.eql('beforeproxy');
        spy.calledOnce.should.be.true;
        done();
      });
  });

  it('should response from local', function(done) {
    var spy = sinon.spy();
    new File('a.js', options)
      .on('found', spy)
      .end(function(err, data) {
        if (err) return done(err);
        data.data.toString().should.be.eql('define("a", function(){});');
        spy.calledOnce.should.be.true;
        spy.calledWith(join(options.directory, 'a.js')).should.be.true;
        done();
      });
  });

  it('should response 404 from local', function(done) {
    var options = {
      directory: fixture,
      cache: false,
    };
    var spy = sinon.spy();
    new File('c.js', options)
      .on('not found', spy)
      .end(function(err) {
        err.should.be.an.instanceof(Error);
        spy.calledOnce.should.be.true;
        spy.calledWith(join(options.directory, 'c.js')).should.be.true;
        done();
      });
  });

  it('should response from remote', function(done) {
    var seajs = fs.readFileSync(join(__dirname, './fixture/sea.js')).toString();
    var spy1 = sinon.spy();
    var spy2 = sinon.spy();
    new File('seajs/seajs/2.1.1/sea.js', options)
      .on('not found', spy1)
      .on('found', spy2)
      .end(function(err, data) {
        if (err) return done(err);
        data.data.toString().should.be.eql(seajs);
        spy1.calledOnce.should.be.true;
        spy2.calledOnce.should.be.true;
        spy1.calledWith(join(options.directory, 'seajs/seajs/2.1.1/sea.js')).should.be.true;
        spy2.calledWith('http://static.alipayobjects.com/seajs/seajs/2.1.1/sea.js').should.be.true;
        done();
      });
  });

  it('should response from remote with multi url', function(done) {
    var options = {
      directory: fixture,
      proxy: [
        'http://i.alipayobjects-xxx.com',
        'http://static.alipayobjects.com',
      ],
      cache: false,
    };
    var seajs = fs.readFileSync(join(__dirname, './fixture/sea.js')).toString();
    var spy1 = sinon.spy();
    var spy2 = sinon.spy();
    new File('seajs/seajs/2.1.1/sea.js', options)
      .on('not found', spy1)
      .on('found', spy2)
      .end(function(err, data) {
        if (err) return done(err);
        data.data.toString().should.be.eql(seajs);
        spy1.calledTwice.should.be.true;
        spy2.calledOnce.should.be.true;
        spy1.calledWith(join(options.directory, 'seajs/seajs/2.1.1/sea.js')).should.be.true;
        spy2.calledWith('http://static.alipayobjects.com/seajs/seajs/2.1.1/sea.js').should.be.true;
        done();
      });
  });

  it('should response when proxy is not only host', function(done) {
    var options = {
      directory: fixture,
      proxy: 'http://static.alipayobjects.com/seajs',
      cache: false,
    };
    var seajs = fs.readFileSync(join(__dirname, './fixture/sea.js')).toString();
    var spy1 = sinon.spy();
    var spy2 = sinon.spy();
    new File('seajs/2.1.1/sea.js', options)
      .on('not found', spy1)
      .on('found', spy2)
      .end(function(err, data) {
        if (err) return done(err);
        data.data.toString().should.be.eql(seajs);
        spy1.calledOnce.should.be.true;
        spy2.calledOnce.should.be.true;
        spy1.calledWith(join(options.directory, 'seajs/2.1.1/sea.js')).should.be.true;
        spy2.calledWith('http://static.alipayobjects.com/seajs/seajs/2.1.1/sea.js').should.be.true;
        done();
      });
  });

  it('should response 404 from remote', function(done) {
    var spy = sinon.spy();
    new File('seajs/seajs/2.1.1/not-exist.js', options)
      .on('not found', spy)
      .end(function(err) {
        err.should.be.an.instanceof(Error);
        spy.calledTwice.should.be.true;
        spy.calledWith(join(options.directory, 'seajs/seajs/2.1.1/not-exist.js')).should.be.true;
        spy.calledWith('http://static.alipayobjects.com/seajs/seajs/2.1.1/not-exist.js').should.be.true;
        done();
      });
  });

  it('should response from remote and cached', function(done) {
    options.cache = true;
    var file = 'seajs/1.3.1/sea.js';
    var cached = join(__dirname, 'fixture', file);
    fs.existsSync(cached) && fs.unlinkSync(cached);

    var spy1 = sinon.spy();
    var spy2 = sinon.spy();
    var spy3 = sinon.spy();
    new File(file, options)
      .on('not found', spy1)
      .on('found', spy2)
      .on('cached', spy3)
      .end(function(err) {
        if (err) return done(err);
        spy1.calledOnce.should.be.true;
        spy2.calledOnce.should.be.true;
        spy3.calledOnce.should.be.true;
        spy1.calledWith(join(options.directory, file)).should.be.true;
        spy2.calledWith(url.resolve('http://static.alipayobjects.com', file)).should.be.true;
        spy3.calledWith(join(options.directory, file)).should.be.true;
        fs.existsSync(cached).should.be.true;
        fs.unlinkSync(cached);
        done();
      });
  });

  it('should get ttf', function(done) {
    var len = fs.readFileSync(join(fixture, 'a.ttf')).length;
    var spy = sinon.spy();
    new File('a.ttf', options)
      .on('found', spy)
      .end(function(err, data) {
        if (err) return done(err);
        spy.calledOnce.should.be.true;
        new Buffer(data.data).length.should.eql(len);
        spy.calledWith(join(options.directory, 'a.ttf')).should.be.true;
        done();
      });
  });

});


test/index.js
'use strict';

var fs = require('fs');
var path = require('path');
var connect = require('connect');
var request = require('supertest');
require('should');
var sinon = require('sinon');

var combo = require('..');

describe('Combo', function() {

  describe('local', function() {
    var options = {
      directory: path.join(__dirname, './fixture'),
    };

    it('should combo different directory', function(done) {
      var app = createServer(options);
      request(app)
        .get('/??a.js,b.js,c/d.js')
        .expect('Content-Type', 'application/javascript')
        .expect('define("a", function(){});define("b", function(){});define("d", function(){});')
        .end(done);
    });

    it('should combo same directory', function(done) {
      var app = createServer(options);
      request(app)
        .get('/c??d.js,e.js')
        .expect('Content-Type', 'application/javascript')
        .expect('define("d", function(){});define("e", function(){});')
        .end(done);
    });

    it('should return 404 when not found', function(done) {
      var app = createServer(options);
      request(app)
        .get('/??a.js,c.js')
        .expect(404, done);
    });

    it('should combo one file', function(done) {
      var app = createServer(options);
      request(app)
        .get('/??a.js')
        .expect('Content-Type', 'application/javascript')
        .expect('define("a", function(){});')
        .end(done);
    });

  });

  describe('remote', function() {
    var options = {
      directory: path.join(__dirname, './fixture'),
      proxy: 'http://static.alipayobjects.com',
    };

    it('should combo different directory', function(done) {
      var app = createServer(options);
      var widget = fs.readFileSync(path.join(__dirname, './fixture/widget.js')).toString();
      request(app)
        .get('/??a.js,arale/widget/1.0.0/widget.js')
        .expect('define("a", function(){});' + widget)
        .end(done);
    });

    it('should 404 when not found', function(done) {
      var app = createServer(options);
      request(app)
        .get('/??a.js,not-exist.js')
        .expect(404, done);
    });
  });

  describe('headers', function() {
    var options = {
      directory: '',
      proxy: 'https://cdnjs.cloudflare.com',
      static: true,
    };

    it('should get header from origin url', function(done) {
      var app = createServer(options);
      request(app)
        .get('/ajax/libs/jquery/3.0.0/core.js')
        .expect('access-control-allow-origin', '*')
        .expect('cache-control', 'public, max-age=30672000')
        .end(done);
    });

    it('should get header from origin url if combo', function(done) {
      var app = createServer(options);
      request(app)
        .get('/??ajax/libs/jquery/3.0.0/core.js')
        .expect('access-control-allow-origin', '*')
        .expect('cache-control', 'public, max-age=30672000')
        .end(done);
    });

    it('should not get header from origin url if enable cache', function(done) {
      options.cache = true;
      options.directory = path.join(__dirname, './fixture');
      var app = createServer(options);
      request(app)
        .get('/ajax/libs/jquery/3.0.0/core.js')
        .expect(function(res) {
          Object.keys(res.headers).should.not.have.property('access-control-allow-origin', 'cache-control');
        })
        .end(done);
    });
  });

  describe('static', function() {
    var options = {
      directory: path.join(__dirname, './fixture'),
      static: true,
    };

    it('should return right', function(done) {
      var app = createServer(options);
      request(app)
        .get('/a.js')
        .expect('define("a", function(){});', done);
    });

    it('should return 404 when not found', function(done) {
      var app = createServer(options);
      request(app)
        .get('/not-exist.js')
        .expect(404, done);
    });

    it('should return right with query', function(done) {
      var app = createServer(options);
      request(app)
        .get('/a.js?t=111')
        .expect('define("a", function(){});', done);
    });
  });

  describe('content type', function() {
    it('static', function(done) {
      var options = {
        directory: path.join(__dirname, './fixture'),
        static: true,
      };
      var app = createServer(options);
      request(app)
        .get('/a.css')
        .expect('Content-Type', 'text/css')
        .end(done);
    });

    it('combo', function(done) {
      var options = {
        directory: path.join(__dirname, './fixture'),
      };
      var app = createServer(options);
      request(app)
        .get('/??a.js,b.js')
        .expect('Content-Type', 'application/javascript')
        .end(done);
    });

    it('400', function(done) {
      var options = {
        directory: path.join(__dirname, './fixture'),
      };
      var app = createServer(options);
      request(app)
        .get('/??a.js,b.css')
        .expect(400, done);
    });
  });


  it('url parse with query', function(done) {
    var options = {
      directory: path.join(__dirname, './fixture'),
    };
    var app = createServer(options);
    request(app)
      .get('/??a.js?123,b.js?456&input_encoding=utf-8')
      .expect('define("a", function(){});define("b", function(){});', done);
  });

  it('should show log', function(done) {
    var log = console.log;
    var spy = sinon.spy();
    console.log = function() {
      spy.apply(null, arguments);
    };
    var options = {
      directory: path.join(__dirname, './fixture'),
      proxy: 'http://ajax.googleapis.com',
      cache: true,
      log: true,
    };
    var app = createServer(options);
    request(app)
      .get('/??a.js,ajax/libs/jquery/1.5.1/jquery.min.js')
      .end(function() {
        spy.calledWithMatch(/>> Found .*\/test\/fixture\/a.js/)
          .should.be.true;
        console.log = log;
        done();
      });
  });

  it('should run next middleware when do not match combo', function(done) {
    var app = createServer();
    request(app)
      .get('/?a.js,b.js')
      .expect('not combo', done);
  });

  it('should return 400 when different ext', function(done) {
    var app = createServer();
    request(app)
      .get('/??a.js,a.css')
      .expect(400, done);
  });

  it('directory function', function(done) {
    var options = {
      static: true,
      directory: function(req) {
        var dir = req.url.split('?')[1].split('=')[1];
        return path.join(__dirname, 'fixture', dir);
      },
    };
    var app = createServer(options);
    request(app)
      .get('/d.js?dir=c')
      .expect('define("d", function(){});', done);
  });

  it('proxy function', function(done) {
    var options = {
      directory: path.join(__dirname, './fixture'),
      proxy: function(req) {
        var key = req.headers.host.split('.')[0];
        return [ 'http://static.alipayobjects.com/' + key ];
      },
    };
    var app = createServer(options);
    var widget = fs.readFileSync(path.join(__dirname, './fixture/widget.js')).toString();
    request(app)
      .get('??/widget/1.0.0/widget.js')
      .set('host', 'arale.alipay.com')
      .expect(widget)
      .end(done);
  });
});

function createServer(options) {
  if (!options) {
    options = {
      directory: path.join(__dirname, './fixture'),
    };
  }
  return connect()
    .use(combo(options))
    .use(function(req, res) {
      res.end('not combo');
    });
}


test/fixture/a.js
define("a", function(){});

lib/file.js
'use strict';

var fs = require('fs');
var path = require('path');
var url = require('url');
var events = require('events');
var util = require('util');
var request = require('urllib').request;
var mkdirp = require('mkdirp');

module.exports = File;

function File(urlPath, options) {
  urlPath = url.parse(urlPath).pathname;
  this.options = options;
  this.urlPath = urlPath;
  this.localPath = path.join(options.directory, urlPath);
  if (!options.proxy) {
    options.proxy = [];
  }
  if (!Array.isArray(options.proxy)) {
    options.proxy = [options.proxy];
  }
  urlPath = '/' + urlPath.replace(/^\//, '');
  this.remotePaths = options.proxy.map(function(host) {
    if (/\/$/.test(host)) {
      host = host.substring(0, host.length - 1);
    }
    return host + urlPath;
  });
}

util.inherits(File, events.EventEmitter);

File.prototype.end = function(cb) {
  var that = this;
  this.getLocal(function(err, data) {
    if (!err) {
      // find file from directory
      that.emit('found', that.localPath);
      cb(null, {
        headers: null,
        data: data,
      });
    } else {
      that.emit('not found', that.localPath);
      if (that.options.beforeProxy) {
        that.emit('before proxy', that.localPath);
        that.options.beforeProxy.call(that, that.urlPath, cb, requestProxy);
      } else {
        requestProxy();
      }
    }

    function requestProxy() {
      if (that.remotePaths.length > 0) {
        // find file from remote server
        that.getRemote(cb);
      } else {
        // not found
        cb(err);
      }
    }
  });
  return this;
};

File.prototype.getLocal = function(cb) {
  fs.readFile(this.localPath, cb);
  return this;
};

File.prototype.getRemote = function(cb) {
  var that = this;
  mapRemoteUrl(this.remotePaths, this.request.bind(this), function(err, data) {
    if (err) {
      return cb(err);
    }
    if (that.options.cache) {
      mkdirp(path.dirname(that.localPath), function(err) {
        if (err) return cb(err);
        fs.writeFile(that.localPath, data.data, function(err) {
          !err && that.emit('cached', that.localPath);
          cb(err, data);
        });
      });
    } else {
      cb(null, data);
    }
  });
  return this;
};

File.prototype.request = function(url, cb) {
  var that = this;
  var opt = {
    gzip: true,
    followRedirect: false,
    rejectUnauthorized: false,
  };
  request(url, opt, function(err, data, res) {
    if (err) {
      that.emit('not found', url);
      return cb(err);
    }

    if (res.statusCode >= 300) {
      that.emit('not found', url);
      return cb(new Error(res.statusCode));
    }

    that.emit('found', url);
    cb(null, {
      headers: shouldGetHeader(that.options) ? getOriginalHeader(res.headers) : null,
      data: data,
    });
  });
};

// if use combo as a pure proxy server, it can get headers from upstream
function shouldGetHeader(options) {
  return !options.cache && options.proxy && options.directory === '';
}

function getOriginalHeader(headers) {
  var allowedHeaders = ['access-control-allow-origin', 'cache-control'];
  return allowedHeaders.reduce(function(pre, curr) {
    if (headers[curr]) {
      pre[curr] = headers[curr];
    }
    return pre;
  }, {});
}

function mapRemoteUrl(arr, fn, cb) {
  arr = arr.slice();

  series(arr, fn, cb);

  function series(arr, fn, cb) {
    if (arr.length === 0) {
      return cb(new Error('Can not get data from urls'));
    }
    var url = arr.shift();
    fn(url, function(err, data) {
      // 如果出错继续调用下一个
      if (err) {
        return series(arr, fn, cb);
      }
      cb(null, data);
    });
  }
}


index.js
'use strict';

var path = require('path');
var debug = require('debug')('connect-combo');
var async = require('async');
var mime = require('mime');
var extend = require('extend');
var parse = require('url').parse;
var File = require('./lib/file');

module.exports = combo;
module.exports.File = File;

var defaults = {
  // local directory
  directory: process.cwd(),

  beforeProxy: null,

  // remote server
  proxy: '',

  // cache remote file
  cache: false,

  // show log
  log: false,

  // serve as normal static server, also support proxy,
  // or you can use `connect.static` without proxy
  static: false,
};

function combo(options) {

  options = extend({}, defaults, options);

  var directory = options.directory;
  var getDir = typeof options.directory === 'string' ?
    function() { return directory; } : options.directory;

  var proxy = options.proxy;
  var getProxy = typeof options.proxy !== 'function' ?
    function() { return proxy; } : options.proxy;

  return function(req, res, next) {
    var opt = extend({}, options, {
      directory: getDir(req),
      proxy: getProxy(req),
    });

    debug('Request %s', req.url);
    var files = normalize(req.url);
    debug('Split %s', files);
    var exts = getExt(files);
    debug('Extension %s', exts);

    var isCombo = decodeURIComponent(req.url).indexOf('??') > -1;
    if (isCombo && files.length !== 0) {
      log('Request ' + req.url, opt);

      if (exts.length !== 1) {
        res.writeHead(400, {'Content-Type': 'text/html'});
        res.end('400 Bad Request');
      } else {
        async.map(files, function(item, done) {
          var f = new File(item, opt).end(done);
          logFile(f, opt);
        }, function(err, results){
          if (err) {
            res.writeHead(404, {'Content-Type': 'text/html'});
            res.end('404 Not Found');
          } else {
            var headers = {
              'Content-Type': mime.lookup(exts[0]),
              'Date': new Date().toUTCString(),
            };
            // get the header from first combo file
            if (results[0].headers !== null) {
              headers = extend({}, headers, results[0].headers);
            }
            res.writeHead(200, headers);

            var tmpres = results.map(function(r) {
              return r.data;
            });
            res.end(Buffer.concat(tmpres));
          }
        });
      }
    } else if (opt.static) {

      log('Request ' + req.url, opt);

      var f = new File(files[0], opt).end(function(err, data) {
        if (err) {
          res.writeHead(404, {'Content-Type': 'text/html'});
          res.end('404 Not Found');
        } else {
          var headers = {
            'Content-Type': mime.lookup(exts[0]),
            'Date': new Date().toUTCString(),
          };
          // get the header from first combo file
          if (data.headers !== null) {
            headers = extend({}, headers, data.headers);
          }
          res.writeHead(200, headers);
          res.end(data.data);
        }
      });
      logFile(f, opt);

    } else {
      // next middleware
      next();
    }
  };
}

// '/a??b.js,c/d.js' => ['a/b.js', 'a/c/d.js']
function normalize(url) {
  url = parse(url);
  var comboPath = extractComboPath(url.query);
  if (comboPath) {
    var base = url.pathname;
    return comboPath
      .split(',')
      .map(function(item) {
        item = parse(item).pathname;
        return path.join(base, item);
      });
  } else {
    return [url.pathname];
  }
}

// /??a.js,b.js&c=d => a.js,b.js
function extractComboPath(query) {
  if (!query) { return; }
  var ret;
  query.split('&')
    .forEach(function(item) {
      item = decodeURIComponent(item);
      if (!ret && item.charAt(0) === '?') {
        ret = item.substring(1).split('=')[0];
      }
    });
  return ret;
}

function logFile(file, options) {
  file
    .on('found', function(str) {
      log('Found ' + str, options);
    })
    .on('not found', function(str) {
      log('Not Found ' + str, options);
    })
    .on('cached', function(str) {
      log('Cached ' + str, options);
    });
}

function log(str, options) {
  options.log && console.log('>> ' + str);
}

// get file extension
// support single file and combo file
function getExt(files) {
  return files.map(function(file) {
    return path.extname(file);
  }).filter(function(item, index, arr) {
    return index === arr.indexOf(item);
  });
}
